// orama/es/ChartBackground/BottomLabel.js
// Copyright 2018 Kensho Technologies, LLC.
import * as React from 'react';
import PropTypes from 'prop-types';
import DEFAULT_THEME from '../defaultTheme'; // component that positions and styles the bottom label of the `Chart` component

export default function BottomLabel(props) {
  var style = {
    bottom: 0,
    color: props.theme.textFill,
    fontFamily: props.theme.fontFamily,
    fontSize: props.theme.axisLabelFontSize,
    fontWeight: props.theme.axisLabelFontWeight,
    left: props.plotRect.x,
    overflow: 'hidden',
    position: 'absolute',
    textAlign: 'center',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    width: props.plotRect.width
  };
  return React.createElement("div", {
    style: style
  }, props.text);
}
BottomLabel.propTypes = process.env.NODE_ENV !== "production" ? {
  plotRect: PropTypes.object,
  text: PropTypes.string,
  theme: PropTypes.object
} : {};
BottomLabel.defaultProps = {
  plotRect: {
    x: 0,
    y: 0,
    width: 0,
    height: 0
  },
  text: '',
  theme: DEFAULT_THEME
};
