// orama/lib/Chart/chartWidthHOC.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/builtin/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/builtin/interopRequireDefault");

exports.__esModule = true;
exports.default = chartWidthHOC;

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/builtin/assertThisInitialized"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/builtin/inheritsLoose"));

var _throttle2 = _interopRequireDefault(require("lodash/throttle"));

var React = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

// Copyright 2018 Kensho Technologies, LLC.
// this component wraps <BaseComponent /> and adds a width prop when it's not present
function chartWidthHOC(BaseComponent) {
  var _class, _temp2;

  return _temp2 = _class =
  /*#__PURE__*/
  function (_React$Component) {
    (0, _inheritsLoose2.default)(ChartWidthHOC, _React$Component);

    function ChartWidthHOC() {
      var _temp, _this;

      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      return (_temp = _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this, _this.state = {}, _this.handleResize = (0, _throttle2.default)(function () {
        return _this.updateWidth();
      }, 500), _this.handleRef = function (divNode) {
        _this.divNode = divNode;
      }, _temp) || (0, _assertThisInitialized2.default)(_this);
    }

    var _proto = ChartWidthHOC.prototype;

    _proto.componentDidMount = function componentDidMount() {
      window.addEventListener('resize', this.handleResize);
      this.updateWidth();
    };

    _proto.componentWillReceiveProps = function componentWillReceiveProps() {
      this.updateWidth();
    };

    _proto.componentWillUnmount = function componentWillUnmount() {
      window.removeEventListener('resize', this.handleResize);
    };

    _proto.updateWidth = function updateWidth() {
      if (this.divNode && !this.props.width) {
        var width = this.divNode.clientWidth;

        if (this.state.width !== width) {
          this.setState({
            width: width
          });
        }
      }
    };

    _proto.render = function render() {
      return React.createElement("div", {
        ref: this.handleRef
      }, React.createElement(BaseComponent, Object.assign({}, this.state, this.props)));
    };

    return ChartWidthHOC;
  }(React.Component), _class.propTypes = {
    width: _propTypes.default.number
  }, _class.displayName = 'ChartWidthHOC', _temp2;
}
