// orama/lib/Chart/ChartRender.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/builtin/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/builtin/interopRequireDefault");

exports.__esModule = true;
exports.default = ChartRender;

var _map2 = _interopRequireDefault(require("lodash/map"));

var React = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _basicRender = _interopRequireDefault(require("../CanvasRender/basicRender"));

var _CanvasRender = _interopRequireDefault(require("../CanvasRender"));

var _defaultTheme = _interopRequireDefault(require("../defaultTheme"));

// Copyright 2018 Kensho Technologies, LLC.
// used inside <Chart />
function ChartRender(props) {
  return React.createElement("div", null, (0, _map2.default)(props.renderLayers, function (renderLayer, i) {
    return React.createElement(_CanvasRender.default // basicRender
    , {
      clip: true,
      height: props.rootProps.height,
      key: i,
      layerProps: renderLayer.layerProps,
      plotRect: props.rootProps.plotRect,
      render: _basicRender.default,
      renderData: renderLayer.renderData,
      theme: props.theme,
      width: props.rootProps.width
    });
  }));
}

ChartRender.propTypes = process.env.NODE_ENV !== "production" ? {
  renderLayers: _propTypes.default.array,
  rootProps: _propTypes.default.object,
  theme: _propTypes.default.object
} : {};
ChartRender.defaultProps = {
  theme: _defaultTheme.default
};
