// orama/es/utilComponents/Portal.js
import _inheritsLoose from "@babel/runtime/helpers/builtin/inheritsLoose";
// Copyright 2018 Kensho Technologies, LLC.
import * as React from 'react';
import { unmountComponentAtNode, unstable_renderSubtreeIntoContainer as render } from 'react-dom';

var Portal =
/*#__PURE__*/
function (_React$Component) {
  _inheritsLoose(Portal, _React$Component);

  function Portal() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = Portal.prototype;

  _proto.componentDidMount = function componentDidMount() {
    this.container = document.createElement('div');
    this.container.position = 'absolute';
    this.container.top = 0;
    this.container.left = 0;
    this.container.right = 0;
    document.body.appendChild(this.container);
    this.componentDidUpdate();
  };

  _proto.componentDidUpdate = function componentDidUpdate() {
    render(this, React.createElement("div", this.props), this.container);
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    unmountComponentAtNode(this.container);
    document.body.removeChild(this.container);
  };

  _proto.render = function render() {
    return null;
  };

  return Portal;
}(React.Component);

export { Portal as default };
