// orama/es/chartCore/defaults.js
// Copyright 2018 Kensho Technologies, LLC.
export var ACCESSORS_NAMES = ['x', 'x0', 'x1', 'x2', 'y', 'y0', 'y1', 'y2', 'radius', 'fill', 'stroke', 'alpha', 'lineWidth', 'lineDash', 'hoverStroke', 'hoverFill'];
export var ACCESSORS_NAMES_NON_SCALABLE = ['fillAlpha', 'strokeAlpha', 'hoverAlpha', 'hoverLineWidth', 'text', 'rotate', 'textAlign', 'textBaseline', 'textSnap', 'xOffset', 'yOffset', 'font', 'fontSize'];
export var ACCESSORS_GROUPS = {
  x: ['x', 'x0', 'x1', 'x2'],
  y: ['y', 'y0', 'y1', 'y2']
};
export var ACCESSORS_TOOLTIP_ORDER = {
  y: 1,
  y0: 2,
  y1: 3,
  y2: 4,
  x: 5,
  x0: 6,
  x1: 7,
  x2: 8,
  radius: 9,
  fill: 10,
  stroke: 11,
  lineWidth: 12
};
export var JS_TO_VIS_TYPE = {
  string: 'ordinal',
  number: 'linear',
  date: 'time'
};
export var WIDTH = 500;
export var HEIGHT = 500;
export var PROPORTION = 0.5;
export var BACKGROUND_OFFSET = 15;
export var RANGE = [0, 1];
export var DOMAIN = [0, 1];
export var TYPE = 'linear';
export var NICE = false;
export var TICK_COUNT = 1;
export var TICK_X_SPACE = 150;
export var TICK_Y_SPACE = 75;
export var SHOW_TICKS = true;
export var SHOW_LABELS = true;
export var AXIS_LABEL_OFFSET = function AXIS_LABEL_OFFSET(theme) {
  return theme.axisLabelFontSize * (theme.lineHeight - 0.75);
};
export var AXIS_TICK_OFFSET = function AXIS_TICK_OFFSET(theme) {
  return theme.axisTickFontSize * (theme.lineHeight - 0.75);
};
export var PLOT_RECT = {
  x: 10,
  y: 10,
  width: 480,
  height: 480
};
export var MARGIN_RIGHT = 15;
