// orama/es/chartCore/getForProps.js
import _reduce from "lodash/reduce";
import { getType, getDomain, getRange, getTickCount, getScale, getTickFormat } from './getForKey';
/*
Functions to be used on the Chart props transformation flow.
The transformation flow starts with the <Chart /> props and successively adds the variables needed for plotting, the transformed props are used for generating render data.
*/

export { default as getDimArrays } from './getDimArrays';
export { getPlotRect } from './getPlotRect';
export var getForProps = function getForProps(value, getFunc) {
  return function (props) {
    return _reduce(props.groupedKeys, function (acc, key) {
      var _Object$assign;

      if (props["" + key + value]) return acc;
      return Object.assign({}, acc, (_Object$assign = {}, _Object$assign["" + key + value] = getFunc(props, key), _Object$assign));
    }, {});
  };
};
export var getTypes = getForProps('Type', getType);
export var getDomains = getForProps('Domain', getDomain);
export var getRanges = getForProps('Range', getRange);
export var getTickCounts = getForProps('TickCount', getTickCount);
export var getScales = getForProps('Scale', getScale);
export var getTickFormatters = getForProps('TickFormat', getTickFormat);
