// orama/es/chartCore/rerunChecks.js
import _some from "lodash/some";

function shallowEqual(objA, objB) {
  if (objA === objB) {
    return true;
  }

  if (typeof objA !== 'object' || objA === null || typeof objB !== 'object' || objB === null) {
    return false;
  }

  var keysA = Object.keys(objA);
  var keysB = Object.keys(objB);

  if (keysA.length !== keysB.length) {
    return false;
  } // Test for A's keys different from B.


  var bHasOwnProperty = Object.prototype.hasOwnProperty.bind(objB);

  for (var i = 0; i < keysA.length; i += 1) {
    if (!bHasOwnProperty(keysA[i]) || objA[keysA[i]] !== objB[keysA[i]]) {
      return false;
    }
  }

  return true;
}

export var getRerunCheckForNames = function getRerunCheckForNames(keyNames, rootNames) {
  return function (props, prevProps) {
    var checkRootNames = _some(rootNames, function (name) {
      return props["" + name] !== prevProps["" + name];
    });

    if (checkRootNames) return true;
    if (!keyNames) return false;
    return _some(props.groupedKeys, function (key) {
      return _some(keyNames, function (name) {
        return props["" + key + name] !== prevProps["" + key + name];
      });
    });
  };
};
export var rerunCheckGetTypes = getRerunCheckForNames(['Type', 'Array']);
export var rerunCheckGetDomains = getRerunCheckForNames(['Domain', 'Array', 'Type', 'ZeroBased']);
export var rerunCheckGetPlotRect = getRerunCheckForNames(['Array', 'ShowTicks', 'ShowLabel'], ['width', 'height', 'margin', 'proportion']);
export var rerunCheckGetRanges = getRerunCheckForNames(['Range', 'Type', 'Array'], ['plotRect']);
export var rerunCheckGetTickCounts = getRerunCheckForNames(['TickCount', 'Range', 'TickSpace']);
export var rerunCheckGetScales = getRerunCheckForNames(['Type', 'Domain', 'Range', 'TickCount', 'Nice']);
export var rerunCheckGetRenderLayers = getRerunCheckForNames(['Array', 'Domain', 'Range', 'Scale'], ['data']); // check change for root and layers on: accessors, data, and skipExtractArrays

export function rerunCheckGetDimArrays(props, prevProps) {
  // now there are layers
  if (!prevProps.layers && props.layers) return true; // for each layer

  var layerCheck = _some(props.layers, function (layer, i) {
    // should skip
    if (layer.skipExtractArrays) return false;
    var prevLayer = prevProps.layers[i]; // new layer

    if (!prevLayer) return true; // new data on the layer

    if (layer.data !== prevLayer.data) return true; // new accessors on the layer

    return !shallowEqual(layer.localAccessors, prevLayer.localAccessors);
  });

  if (layerCheck) return true;
  return false;
}
