// orama/lib/ChartBackground/BottomLabel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/builtin/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/builtin/interopRequireWildcard");

exports.__esModule = true;
exports.default = BottomLabel;

var React = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _defaultTheme = _interopRequireDefault(require("../defaultTheme"));

// Copyright 2018 Kensho Technologies, LLC.
// component that positions and styles the bottom label of the `Chart` component
function BottomLabel(props) {
  var style = {
    bottom: 0,
    color: props.theme.textFill,
    fontFamily: props.theme.fontFamily,
    fontSize: props.theme.axisLabelFontSize,
    fontWeight: props.theme.axisLabelFontWeight,
    left: props.plotRect.x,
    overflow: 'hidden',
    position: 'absolute',
    textAlign: 'center',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    width: props.plotRect.width
  };
  return React.createElement("div", {
    style: style
  }, props.text);
}

BottomLabel.propTypes = process.env.NODE_ENV !== "production" ? {
  plotRect: _propTypes.default.object,
  text: _propTypes.default.string,
  theme: _propTypes.default.object
} : {};
BottomLabel.defaultProps = {
  plotRect: {
    x: 0,
    y: 0,
    width: 0,
    height: 0
  },
  text: '',
  theme: _defaultTheme.default
};
