// orama/lib/ChartBackground/LeftLabel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/builtin/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/builtin/interopRequireWildcard");

exports.__esModule = true;
exports.default = LeftLabel;

var React = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _defaultTheme = _interopRequireDefault(require("../defaultTheme"));

// Copyright 2018 Kensho Technologies, LLC.
// component that positions and styles the left label of the `Chart` component
function LeftLabel(props) {
  var style = {
    color: props.theme.textFill,
    fontFamily: props.theme.fontFamily,
    fontSize: props.theme.axisLabelFontSize,
    fontWeight: props.theme.axisLabelFontWeight,
    left: 0,
    overflow: 'hidden',
    position: 'absolute',
    textAlign: 'center',
    textOverflow: 'ellipsis',
    top: props.plotRect.y,
    transform: 'translate(-100%) rotate(-90deg)',
    transformOrigin: '100% 0',
    whiteSpace: 'nowrap',
    width: props.plotRect.height
  };
  return React.createElement("div", {
    style: style
  }, props.text);
}

LeftLabel.propTypes = process.env.NODE_ENV !== "production" ? {
  plotRect: _propTypes.default.object,
  text: _propTypes.default.string,
  theme: _propTypes.default.object
} : {};
LeftLabel.defaultProps = {
  plotRect: {
    x: 0,
    y: 0,
    width: 0,
    height: 0
  },
  text: '',
  theme: _defaultTheme.default
};
