// orama/es/Layer/index.js
// Copyright 2018 Kensho Technologies, LLC.
import areas from './areas';
import bars from './bars';
import brushes from './brushes';
import guides from './guides';
import lines from './lines';
import points from './points';
import ranges from './ranges';
import text from './text';
import scatterplotLabels from './scatterplotLabels';
export var Layer = function Layer() {
  return null;
};
export var Areas = function Areas() {
  return null;
};
Areas.plot = areas;
export var Bars = function Bars() {
  return null;
};
Bars.plot = bars;
export var Brushes = function Brushes() {
  return null;
};
Brushes.plot = brushes;
export var Guides = function Guides() {
  return null;
};
Guides.plot = guides;
export var Lines = function Lines() {
  return null;
};
Lines.plot = lines;
export var Points = function Points() {
  return null;
};
Points.plot = points;
export var Ranges = function Ranges() {
  return null;
};
Ranges.plot = ranges;
export var Text = function Text() {
  return null;
};
Text.plot = text;
export var ScatterplotLabels = function ScatterplotLabels() {
  return null;
};
ScatterplotLabels.plot = scatterplotLabels;
