// orama/es/Chart/ChartRender.js
import _map from "lodash/map";
// Copyright 2018 Kensho Technologies, LLC.
import * as React from 'react';
import PropTypes from 'prop-types';
import basicRender from '../CanvasRender/basicRender';
import CanvasRender from '../CanvasRender';
import DEFAULT_THEME from '../defaultTheme'; // used inside <Chart />

export default function ChartRender(props) {
  return React.createElement("div", null, _map(props.renderLayers, function (renderLayer, i) {
    return React.createElement(CanvasRender // basicRender
    , {
      clip: true,
      height: props.rootProps.height,
      key: i,
      layerProps: renderLayer.layerProps,
      plotRect: props.rootProps.plotRect,
      render: basicRender,
      renderData: renderLayer.renderData,
      theme: props.theme,
      width: props.rootProps.width
    });
  }));
}
ChartRender.propTypes = process.env.NODE_ENV !== "production" ? {
  renderLayers: PropTypes.array,
  rootProps: PropTypes.object,
  theme: PropTypes.object
} : {};
ChartRender.defaultProps = {
  theme: DEFAULT_THEME
};
