// orama/lib/utilComponents/Portal.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/builtin/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/builtin/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/builtin/inheritsLoose"));

var React = _interopRequireWildcard(require("react"));

var _reactDom = require("react-dom");

// Copyright 2018 Kensho Technologies, LLC.
var Portal =
/*#__PURE__*/
function (_React$Component) {
  (0, _inheritsLoose2.default)(Portal, _React$Component);

  function Portal() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = Portal.prototype;

  _proto.componentDidMount = function componentDidMount() {
    this.container = document.createElement('div');
    this.container.position = 'absolute';
    this.container.top = 0;
    this.container.left = 0;
    this.container.right = 0;
    document.body.appendChild(this.container);
    this.componentDidUpdate();
  };

  _proto.componentDidUpdate = function componentDidUpdate() {
    (0, _reactDom.unstable_renderSubtreeIntoContainer)(this, React.createElement("div", this.props), this.container);
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    (0, _reactDom.unmountComponentAtNode)(this.container);
    document.body.removeChild(this.container);
  };

  _proto.render = function render() {
    return null;
  };

  return Portal;
}(React.Component);

exports.default = Portal;
