// orama/lib/chartCore/defaults.js
"use strict";

exports.__esModule = true;
exports.MARGIN_RIGHT = exports.PLOT_RECT = exports.AXIS_TICK_OFFSET = exports.AXIS_LABEL_OFFSET = exports.SHOW_LABELS = exports.SHOW_TICKS = exports.TICK_Y_SPACE = exports.TICK_X_SPACE = exports.TICK_COUNT = exports.NICE = exports.TYPE = exports.DOMAIN = exports.RANGE = exports.BACKGROUND_OFFSET = exports.PROPORTION = exports.HEIGHT = exports.WIDTH = exports.JS_TO_VIS_TYPE = exports.ACCESSORS_TOOLTIP_ORDER = exports.ACCESSORS_GROUPS = exports.ACCESSORS_NAMES_NON_SCALABLE = exports.ACCESSORS_NAMES = void 0;
// Copyright 2018 Kensho Technologies, LLC.
var ACCESSORS_NAMES = ['x', 'x0', 'x1', 'x2', 'y', 'y0', 'y1', 'y2', 'radius', 'fill', 'stroke', 'alpha', 'lineWidth', 'lineDash', 'hoverStroke', 'hoverFill'];
exports.ACCESSORS_NAMES = ACCESSORS_NAMES;
var ACCESSORS_NAMES_NON_SCALABLE = ['fillAlpha', 'strokeAlpha', 'hoverAlpha', 'hoverLineWidth', 'text', 'rotate', 'textAlign', 'textBaseline', 'textSnap', 'xOffset', 'yOffset', 'font', 'fontSize'];
exports.ACCESSORS_NAMES_NON_SCALABLE = ACCESSORS_NAMES_NON_SCALABLE;
var ACCESSORS_GROUPS = {
  x: ['x', 'x0', 'x1', 'x2'],
  y: ['y', 'y0', 'y1', 'y2']
};
exports.ACCESSORS_GROUPS = ACCESSORS_GROUPS;
var ACCESSORS_TOOLTIP_ORDER = {
  y: 1,
  y0: 2,
  y1: 3,
  y2: 4,
  x: 5,
  x0: 6,
  x1: 7,
  x2: 8,
  radius: 9,
  fill: 10,
  stroke: 11,
  lineWidth: 12
};
exports.ACCESSORS_TOOLTIP_ORDER = ACCESSORS_TOOLTIP_ORDER;
var JS_TO_VIS_TYPE = {
  string: 'ordinal',
  number: 'linear',
  date: 'time'
};
exports.JS_TO_VIS_TYPE = JS_TO_VIS_TYPE;
var WIDTH = 500;
exports.WIDTH = WIDTH;
var HEIGHT = 500;
exports.HEIGHT = HEIGHT;
var PROPORTION = 0.5;
exports.PROPORTION = PROPORTION;
var BACKGROUND_OFFSET = 15;
exports.BACKGROUND_OFFSET = BACKGROUND_OFFSET;
var RANGE = [0, 1];
exports.RANGE = RANGE;
var DOMAIN = [0, 1];
exports.DOMAIN = DOMAIN;
var TYPE = 'linear';
exports.TYPE = TYPE;
var NICE = false;
exports.NICE = NICE;
var TICK_COUNT = 1;
exports.TICK_COUNT = TICK_COUNT;
var TICK_X_SPACE = 150;
exports.TICK_X_SPACE = TICK_X_SPACE;
var TICK_Y_SPACE = 75;
exports.TICK_Y_SPACE = TICK_Y_SPACE;
var SHOW_TICKS = true;
exports.SHOW_TICKS = SHOW_TICKS;
var SHOW_LABELS = true;
exports.SHOW_LABELS = SHOW_LABELS;

var AXIS_LABEL_OFFSET = function AXIS_LABEL_OFFSET(theme) {
  return theme.axisLabelFontSize * (theme.lineHeight - 0.75);
};

exports.AXIS_LABEL_OFFSET = AXIS_LABEL_OFFSET;

var AXIS_TICK_OFFSET = function AXIS_TICK_OFFSET(theme) {
  return theme.axisTickFontSize * (theme.lineHeight - 0.75);
};

exports.AXIS_TICK_OFFSET = AXIS_TICK_OFFSET;
var PLOT_RECT = {
  x: 10,
  y: 10,
  width: 480,
  height: 480
};
exports.PLOT_RECT = PLOT_RECT;
var MARGIN_RIGHT = 15;
exports.MARGIN_RIGHT = MARGIN_RIGHT;
