// orama/lib/Layer/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/builtin/interopRequireDefault");

exports.__esModule = true;
exports.ScatterplotLabels = exports.Text = exports.Ranges = exports.Points = exports.Lines = exports.Guides = exports.Brushes = exports.Bars = exports.Areas = exports.Layer = void 0;

var _areas = _interopRequireDefault(require("./areas"));

var _bars = _interopRequireDefault(require("./bars"));

var _brushes = _interopRequireDefault(require("./brushes"));

var _guides = _interopRequireDefault(require("./guides"));

var _lines = _interopRequireDefault(require("./lines"));

var _points = _interopRequireDefault(require("./points"));

var _ranges = _interopRequireDefault(require("./ranges"));

var _text = _interopRequireDefault(require("./text"));

var _scatterplotLabels = _interopRequireDefault(require("./scatterplotLabels"));

// Copyright 2018 Kensho Technologies, LLC.
var Layer = function Layer() {
  return null;
};

exports.Layer = Layer;

var Areas = function Areas() {
  return null;
};

exports.Areas = Areas;
Areas.plot = _areas.default;

var Bars = function Bars() {
  return null;
};

exports.Bars = Bars;
Bars.plot = _bars.default;

var Brushes = function Brushes() {
  return null;
};

exports.Brushes = Brushes;
Brushes.plot = _brushes.default;

var Guides = function Guides() {
  return null;
};

exports.Guides = Guides;
Guides.plot = _guides.default;

var Lines = function Lines() {
  return null;
};

exports.Lines = Lines;
Lines.plot = _lines.default;

var Points = function Points() {
  return null;
};

exports.Points = Points;
Points.plot = _points.default;

var Ranges = function Ranges() {
  return null;
};

exports.Ranges = Ranges;
Ranges.plot = _ranges.default;

var Text = function Text() {
  return null;
};

exports.Text = Text;
Text.plot = _text.default;

var ScatterplotLabels = function ScatterplotLabels() {
  return null;
};

exports.ScatterplotLabels = ScatterplotLabels;
ScatterplotLabels.plot = _scatterplotLabels.default;
