// orama/es/CanvasInput/index.js
import _assertThisInitialized from "@babel/runtime/helpers/builtin/assertThisInitialized";
import _inheritsLoose from "@babel/runtime/helpers/builtin/inheritsLoose";
// Copyright 2018 Kensho Technologies, LLC.
import * as React from 'react';
import PropTypes from 'prop-types';
import hoverRender from '../CanvasRender/hoverRender';
import TooltipWrapper from '../TooltipWrapper';
import CanvasRender from '../CanvasRender';
import { getDataUnderMouse } from './getDataUnderMouse';
import { getMouseFromEvt, runHoverSolverOn } from './methods';
/*
Usually used inside of <ChartRender />
Get hovered and clicked data on renderData using a <canvas /> element
*/

var CanvasInput =
/*#__PURE__*/
function (_React$Component) {
  _inheritsLoose(CanvasInput, _React$Component);

  function CanvasInput() {
    var _temp, _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return (_temp = _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this, _this.state = {}, _this.handleCanvasRef = function (canvasNode) {
      _this.canvasNode = canvasNode;
    }, _this.handleClick = function (evt) {
      evt.stopPropagation();
      evt.preventDefault();

      if (!_this.state.mouseDrag) {
        var mouse = getMouseFromEvt(evt);
        var solvedData = runHoverSolverOn(getDataUnderMouse(_this.props, mouse, _this.canvasNode));

        _this.props.onUpdate({
          action: 'mouseClick',
          mouse: mouse,
          renderDatum: solvedData.renderDatum,
          hoverRenderData: solvedData.hoverRenderData,
          hoverData: solvedData.hoverData,
          localMouse: solvedData.localMouse,
          layerProps: solvedData.layerProps,
          rootProps: _this.props.rootProps
        });
      }

      _this.setState({
        mouseDrag: false
      });
    }, _this.handleDoubleClick = function () {
      _this.props.onUpdate({
        action: 'mouseDoubleClick'
      });
    }, _this.handleMouseDown = function (evt) {
      var mouse = getMouseFromEvt(evt);
      var solvedData = runHoverSolverOn(getDataUnderMouse(_this.props, mouse, _this.canvasNode));

      _this.props.onUpdate({
        action: 'mouseDown',
        mouse: mouse,
        renderDatum: solvedData.renderDatum,
        hoverRenderData: solvedData.hoverRenderData,
        hoverData: solvedData.hoverData,
        localMouse: solvedData.localMouse,
        layerProps: solvedData.layerProps,
        rootProps: _this.props.rootProps
      });

      _this.setState({
        mouseDown: true,
        mouse: mouse,
        hoverRenderData: solvedData.hoverRenderData,
        hoverData: solvedData.hoverData,
        layerProps: solvedData.layerProps
      });

      _this.lastMouse = mouse;
    }, _this.handleMouseMove = function (evt) {
      var mouse = getMouseFromEvt(evt);
      var solvedData = runHoverSolverOn(getDataUnderMouse(_this.props, mouse, _this.canvasNode));
      var mouseDelta;

      if (_this.lastMouse) {
        mouseDelta = {
          x: _this.lastMouse.x - mouse.x,
          y: _this.lastMouse.y - mouse.y
        };
      }

      _this.props.onUpdate({
        action: _this.state.mouseDrag ? 'mouseDrag' : 'mouseMove',
        mouse: mouse,
        mouseDelta: mouseDelta,
        renderDatum: solvedData.renderDatum,
        hoverRenderData: solvedData.hoverRenderData,
        hoverData: solvedData.hoverData,
        localMouse: solvedData.localMouse,
        layerProps: solvedData.layerProps,
        rootProps: _this.props.rootProps
      });

      _this.setState(function (prevState) {
        return {
          mouseDrag: !!prevState.mouseDown,
          mouse: mouse,
          hoverRenderData: solvedData.hoverRenderData,
          hoverData: solvedData.hoverData,
          layerProps: solvedData.layerProps
        };
      });

      _this.lastMouse = mouse;
      _this.mouseLeave = false;
    }, _this.handleMouseUp = function (evt) {
      evt.stopPropagation();
      evt.preventDefault();
      var mouse = getMouseFromEvt(evt);
      var solvedData = runHoverSolverOn(getDataUnderMouse(_this.props, mouse, _this.canvasNode, evt));

      _this.props.onUpdate({
        action: 'mouseUp',
        mouse: mouse,
        renderDatum: solvedData.renderDatum,
        hoverRenderData: solvedData.hoverRenderData,
        hoverData: solvedData.hoverData,
        localMouse: solvedData.localMouse,
        layerProps: solvedData.layerProps,
        rootProps: _this.props.rootProps
      });

      _this.setState({
        mouseDrag: false,
        mouseDown: false
      });
    }, _this.handleMouseLeave = function () {
      _this.props.onUpdate({
        action: 'mouseLeave'
      });

      _this.setState({
        mouse: undefined,
        hoverRenderData: undefined,
        hoverData: undefined,
        layerProps: undefined
      });

      _this.mouseLeave = true;
    }, _temp) || _assertThisInitialized(_this);
  }

  var _proto = CanvasInput.prototype;

  _proto.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    if (this.state.mouse && !this.mouseLeave) {
      var solvedData = runHoverSolverOn(getDataUnderMouse(nextProps, this.state.mouse, this.canvasNode));
      this.setState({
        renderDatum: solvedData.renderDatum,
        hoverRenderData: solvedData.hoverRenderData,
        hoverData: solvedData.hoverData,
        localMouse: solvedData.localMouse,
        layerProps: solvedData.layerProps
      });
    }
  };

  _proto.componentDidUpdate = function componentDidUpdate(prevProps, prevState) {
    if (this.state.mouseDrag && !prevState.mouseDrag) {
      document.addEventListener('mouseup', this.handleMouseUp);
    } else if (!this.state.mouseDrag && prevState.mouseDrag) {
      document.removeEventListener('mouseup', this.handleMouseUp);
    }
  };

  _proto.render = function render() {
    var _this$props = this.props,
        rootProps = _this$props.rootProps,
        theme = _this$props.theme;
    var _this$state = this.state,
        hoverRenderData = _this$state.hoverRenderData,
        hoverData = _this$state.hoverData,
        layerProps = _this$state.layerProps,
        mouse = _this$state.mouse;
    return React.createElement("div", null, React.createElement(CanvasRender // hoverRender
    , {
      clip: true,
      height: rootProps.height,
      plotRect: rootProps.plotRect,
      render: hoverRender,
      renderData: hoverRenderData,
      theme: theme,
      width: rootProps.width
    }), React.createElement("canvas", {
      height: rootProps.height,
      onClick: this.handleClick,
      onDoubleClick: this.handleDoubleClick,
      onMouseDown: this.handleMouseDown,
      onMouseLeave: this.handleMouseLeave,
      onMouseMove: this.handleMouseMove,
      onMouseUp: this.handleMouseUp,
      onTouchEnd: this.handleMouseLeave,
      onTouchMove: this.handleMouseMove,
      onTouchStart: this.handleMouseDown,
      ref: this.handleCanvasRef,
      style: {
        cursor: 'pointer',
        display: 'block',
        position: 'absolute',
        userSelect: 'none',
        width: rootProps.width,
        height: rootProps.height
      },
      width: rootProps.width
    }), mouse && hoverData ? React.createElement(TooltipWrapper, {
      hoverData: hoverData,
      layerProps: layerProps,
      mouse: mouse,
      theme: theme
    }) : null);
  };

  return CanvasInput;
}(React.Component);

CanvasInput.defaultProps = {
  renderLayers: []
};
export { CanvasInput as default };
CanvasInput.propTypes = process.env.NODE_ENV !== "production" ? {
  onUpdate: PropTypes.func.isRequired,
  renderLayers: PropTypes.array,
  rootProps: PropTypes.object,
  theme: PropTypes.object
} : {};
