// orama/es/utils/State.js
import _assertThisInitialized from "@babel/runtime/helpers/builtin/assertThisInitialized";
import _inheritsLoose from "@babel/runtime/helpers/builtin/inheritsLoose";
import _omit from "lodash/omit";
import _isPlainObject from "lodash/isPlainObject";
import _isFunction from "lodash/isFunction";
// Copyright 2018 Kensho Technologies, LLC.
import * as React from 'react';
import PropTypes from 'prop-types';

var stateHOC =
/*#__PURE__*/
function (_React$PureComponent) {
  _inheritsLoose(stateHOC, _React$PureComponent);

  function stateHOC() {
    var _temp, _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return (_temp = _this = _React$PureComponent.call.apply(_React$PureComponent, [this].concat(args)) || this, _this.state = {}, _this.handleChildUpdate = function (childProps) {
      _this.setState(childProps);
    }, _this.childrenMap = function (child, key) {
      if (key === void 0) {
        key = 0;
      }

      return React.cloneElement(child, Object.assign({
        onUpdate: _this.handleChildUpdate
      }, _omit(_this.props, ['children', 'startWith']), _this.state, {
        setState: _this.handleChildUpdate,
        key: key
      }));
    }, _temp) || _assertThisInitialized(_this);
  }

  var _proto = stateHOC.prototype;

  _proto.componentWillMount = function componentWillMount() {
    if (this.props.startWith && _isFunction(this.props.startWith)) {
      this.props.startWith(Object.assign({
        onUpdate: this.props.onUpdate || this.handleChildUpdate
      }, _omit(this.props, ['children', 'startWith']), this.state, {
        setState: this.handleChildUpdate
      }));
    } else if (this.props.startWith && _isPlainObject(this.props.startWith)) {
      this.setState(this.props.startWith);
    }
  };

  _proto.render = function render() {
    if (React.Children.count(this.props.children) === 0) return React.createElement("span", null);

    if (React.Children.count(this.props.children) > 1) {
      return React.createElement("div", null, React.Children.map(this.props.children, this.childrenMap));
    }

    return this.childrenMap(React.Children.only(this.props.children));
  };

  return stateHOC;
}(React.PureComponent);

stateHOC.defaultProps = {};
export { stateHOC as default };
stateHOC.propTypes = process.env.NODE_ENV !== "production" ? {
  children: PropTypes.node,
  onUpdate: PropTypes.func,
  startWith: PropTypes.oneOfType([PropTypes.func, PropTypes.object])
} : {};
