// orama/lib/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/builtin/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/builtin/interopRequireDefault");

exports.__esModule = true;
exports.chartCore = exports.utils = exports.ScatterplotLabels = exports.Text = exports.Ranges = exports.Points = exports.Lines = exports.Guides = exports.Brushes = exports.Bars = exports.Areas = exports.Layer = exports.getCachedContext = exports.getPath2D = exports.getTimeSeries = exports.Highlight = exports.Brush = exports.Chart = exports.getPlotValues = exports.plotValue = exports.text = exports.ranges = exports.points = exports.lines = exports.guides = exports.brushes = exports.bars = exports.areas = exports.DEFAULT_THEME = void 0;

var _defaultTheme = _interopRequireDefault(require("./defaultTheme"));

exports.DEFAULT_THEME = _defaultTheme.default;

var _dataGeneration = require("./utils/dataGeneration");

exports.getTimeSeries = _dataGeneration.getTimeSeries;

var _path2DUtils = require("./utils/path2DUtils");

exports.getPath2D = _path2DUtils.getPath2D;

var _canvasUtils = require("./utils/canvasUtils");

exports.getCachedContext = _canvasUtils.getCachedContext;

var _utils = _interopRequireWildcard(require("./utils"));

exports.utils = _utils;

var _chartCore = _interopRequireWildcard(require("./chartCore"));

exports.chartCore = _chartCore;

var _areas = _interopRequireDefault(require("./Layer/areas"));

exports.areas = _areas.default;

var _bars = _interopRequireDefault(require("./Layer/bars"));

exports.bars = _bars.default;

var _brushes = _interopRequireDefault(require("./Layer/brushes"));

exports.brushes = _brushes.default;

var _guides = _interopRequireDefault(require("./Layer/guides"));

exports.guides = _guides.default;

var _lines = _interopRequireDefault(require("./Layer/lines"));

exports.lines = _lines.default;

var _points = _interopRequireDefault(require("./Layer/points"));

exports.points = _points.default;

var _ranges = _interopRequireDefault(require("./Layer/ranges"));

exports.ranges = _ranges.default;

var _text = _interopRequireDefault(require("./Layer/text"));

exports.text = _text.default;

var _plotValue = _interopRequireDefault(require("./Layer/plotValue"));

exports.plotValue = _plotValue.default;

var _getPlotValues = _interopRequireDefault(require("./Layer/getPlotValues"));

exports.getPlotValues = _getPlotValues.default;

var _Chart = _interopRequireDefault(require("./Chart"));

exports.Chart = _Chart.default;

var _Layer = require("./Layer");

exports.Layer = _Layer.Layer;
exports.Areas = _Layer.Areas;
exports.Bars = _Layer.Bars;
exports.Brushes = _Layer.Brushes;
exports.Guides = _Layer.Guides;
exports.Lines = _Layer.Lines;
exports.Points = _Layer.Points;
exports.Ranges = _Layer.Ranges;
exports.Text = _Layer.Text;
exports.ScatterplotLabels = _Layer.ScatterplotLabels;

var _Brush = _interopRequireDefault(require("./extensions/Brush"));

exports.Brush = _Brush.default;

var _Highlight = _interopRequireDefault(require("./extensions/Highlight"));

exports.Highlight = _Highlight.default;
