// orama/es/Tooltip.js
import _map from "lodash/map";
// Copyright 2018 Kensho Technologies, LLC.
import * as React from 'react';
import PropTypes from 'prop-types';
import DEFAULT_THEME from './defaultTheme';
import extractTooltipData from './Chart/extractTooltipData';
var MAX_WIDTH = 320;

var getPadding = function getPadding(props) {
  return props.theme.tooltipFontSize / 2;
};

var row = function row(props, d, i) {
  return React.createElement("tr", {
    style: {
      background: i % 2 ? props.theme.tooltipBackgroundFill : props.theme.tooltipEvenBackgroundFill
    },
    key: i
  }, props.showKeys ? React.createElement("td", {
    style: {
      borderRight: "2px solid " + props.theme.tooltipKeyBorderStroke,
      padding: getPadding(props)
    }
  }, d.key) : null, React.createElement("td", {
    // Name
    style: {
      padding: getPadding(props),
      textAlign: 'left',
      verticalAlign: 'top'
    }
  }, d.name), React.createElement("td", {
    // Value
    style: {
      fontFamily: props.theme.fontFamilyMono,
      fontSize: props.theme.tooltipValueFontSize,
      padding: getPadding(props),
      textAlign: 'right',
      verticalAlign: 'top'
    }
  }, d.value));
};

row.propTypes = process.env.NODE_ENV !== "production" ? {
  showKeys: PropTypes.bool,
  theme: PropTypes.object
} : {};

var TooltipInner = function TooltipInner(props) {
  return React.createElement("div", {
    style: {
      background: props.theme.tooltipBackgroundFill,
      boxShadow: "1px 1px 1px " + props.theme.tooltipBoxShadowFill,
      color: props.theme.tooltipTextFill,
      fontFamily: props.theme.fontFamily,
      fontSize: props.theme.tooltipFontSize,
      maxWidth: MAX_WIDTH,
      opacity: 0.96
    }
  }, props.title ? React.createElement("div", {
    style: {
      fontSize: props.theme.tooltipTitleFontSize,
      fontWeight: props.theme.tooltipTitleFontWeight,
      padding: getPadding(props),
      textAlign: 'left',
      verticalAlign: 'top'
    }
  }, props.title) : null, React.createElement("table", {
    style: {
      width: '100%',
      borderCollapse: 'collapse'
    }
  }, React.createElement("tbody", null, _map(props.values, function (d, i) {
    return row(props, d, i);
  }))));
};

TooltipInner.propTypes = process.env.NODE_ENV !== "production" ? {
  theme: PropTypes.object,
  title: PropTypes.string,
  values: PropTypes.array
} : {};
TooltipInner.defaultProps = {
  theme: DEFAULT_THEME,
  showKeys: false
};
export default function Tooltip(props) {
  var tooltipData = extractTooltipData(props.layerProps, props.hoverData);
  return React.createElement(TooltipInner, Object.assign({
    showKeys: props.layerProps.tooltipShowKeys,
    theme: props.theme
  }, tooltipData));
}
Tooltip.propTypes = process.env.NODE_ENV !== "production" ? {
  hoverData: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  layerProps: PropTypes.object,
  theme: PropTypes.object
} : {};
